include/xtensor/xbuilder.hpp

```diff
@@ -202,6 +202,121 @@ namespace xt
+            using xtype = T;
+            using value_type = typename T::value_type;
+
+            diagonal_fn(const T& arr) : m_arr(arr)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(const It& begin, const It& /*end*/) const
+            {
+                return m_arr(*begin, *begin);
+            }
+
+        private:
+            const xtype& m_arr;
+        };
+
+        template <class T>
+        struct diag_fn
+        {
+            using xtype = T;
+            using value_type = typename T::value_type;
+
+            diag_fn(const T& arr) : m_arr(arr)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(const It& begin, const It& /*end*/) const
+            {
+                auto other = begin;
+                other += 1;
+                return *begin == *other ? m_arr(*begin) : value_type(0);
+            }
+
+        private:
+            const T& m_arr;
+        };
+
+        template <class T>
+        struct flipud_fn
+        {
+            using xtype = T;
+            using value_type = typename T::value_type;
+
+            flipud_fn(const T& arr) : m_arr(arr), shape_first(m_arr.shape().front() - 1)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(const It& begin, const It& end) const
+            {
+                xindex idx(begin, end);
+                idx.front() = shape_first - idx.front();
+                return m_arr.element(idx.begin(), idx.end());
+            }
+
+        private:
+            const T& m_arr;
+            const std::size_t shape_first;
+        };
+
+        template <class T>
+        struct fliplr_fn
+        {
+            using xtype = T;
+            using value_type = typename T::value_type;
+
+            fliplr_fn(const T& arr) : m_arr(arr), shape_last(m_arr.shape().back() - 1)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(const It& begin, const It& end) const
+            {
+                xindex idx(begin, end);
+                idx.back() = shape_last - idx.back();
+                return m_arr.element(idx.begin(), idx.end());
+            }
+
+        private:
+            const T& m_arr;
+            const std::size_t shape_last;
+        };
+
+        template <class T, class C>
+        struct tril_fn
+        {
+            using xtype = T;
+            using value_type = typename T::value_type;
+            using signed_idx_type = long int;
+
+            tril_fn(const T& arr, int k, const C& comp) : m_arr(arr), m_k(k), m_comp(comp)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(const It& begin, const It& end) const
+            {
+                // have to cast to signed int otherwise -1 can lead to overflow
+                auto begin_next = begin;
+                begin_next += 1;
```

Comment:
As previous, with a rebase you should now be able to write ``auto begin_next = begin + 1;``